op-chain-ops/deployer/apply.go

```diff
@@ -137,28 +139,43 @@ func ApplyPipeline(
 	intent *state.Intent,
 	st *state.State,
 ) error {
+	progressor := func(curr, total int64) {
+		env.Logger.Info("artifacts download progress", "current", curr, "total", total)
+	}
+
+	artifactsFS, cleanup, err := pipeline.DownloadArtifacts(ctx, intent.ContractArtifactsURL, progressor)
+	if err != nil {
+		return fmt.Errorf("failed to download artifacts: %w", err)
+	}
+	defer func() {
+		if err := cleanup(); err != nil {
+			env.Logger.Warn("failed to clean up artifacts", "err", err)
+		}
+	}()
+
 	pline := []pipelineStage{
 		{"init", pipeline.Init},
 		{"deploy-superchain", pipeline.DeploySuperchain},
 		{"deploy-implementations", pipeline.DeployImplementations},
 	}
 
 	for _, chain := range intent.Chains {
+		chainID := chain.ID
 		pline = append(pline, pipelineStage{
-			fmt.Sprintf("deploy-opchain-%s", chain.ID.Hex()),
-			func(ctx context.Context, env *pipeline.Env, intent *state.Intent, st *state.State) error {
-				return pipeline.DeployOPChain(ctx, env, intent, st, chain.ID)
+			fmt.Sprintf("deploy-opchain-%s", chainID.Hex()),
+			func(ctx context.Context, env *pipeline.Env, artifactsFS foundry.StatDirFs, intent *state.Intent, st *state.State) error {
+				return pipeline.DeployOPChain(ctx, env, artifactsFS, intent, st, chainID)
 			},
 		}, pipelineStage{
-			fmt.Sprintf("generate-l2-genesis-%s", chain.ID.Hex()),
-			func(ctx context.Context, env *pipeline.Env, intent *state.Intent, st *state.State) error {
-				return pipeline.GenerateL2Genesis(ctx, env, intent, st, chain.ID)
+			fmt.Sprintf("generate-l2-genesis-%s", chainID.Hex()),
+			func(ctx context.Context, env *pipeline.Env, artifactsFS foundry.StatDirFs, intent *state.Intent, st *state.State) error {
+				return pipeline.GenerateL2Genesis(ctx, env, artifactsFS, intent, st, chainID)
 			},
 		})
 	}
 
 	for _, stage := range pline {
-		if err := stage.stage(ctx, env, intent, st); err != nil {
+		if err := stage.stage(ctx, env, artifactsFS, intent, st); err != nil {
```

Comment:
Could `stage.stage` be renamed to `stage.execute` to make it a little more obvious what it is doing?